### PR TITLE
Implements Ramer-Douglas-Peucker simplification algorithm

### PR DESCRIPTION
**Pull request by Sellig6792**
# Pull Request Template

## Description

I implemented the Ramer-Douglas-Peucker algorithm. I use it in a project it works very well. This algorithm allow you to reduce the number of point of a polygon and to keep a maximum of precision.

The test doesn't make a real sesnse, since the results changes with the epsilon value.

- https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm 
- https://fr.wikipedia.org/wiki/Algorithme_de_Douglas-Peucker (FR)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I ran bellow commands using the latest version of **rust nightly**. => My code doesn't make issues
- [x] I ran `cargo clippy --all -- -D warnings` just before my last commit and fixed any issue that was found.
- [x] I ran `cargo fmt` just before my last commit.
- [x] I ran `cargo test` just before my last commit and all tests passed.
- [x] I added my algorithm to the corresponding `mod.rs` file within its own folder, and in any parent folder(s).
- [x] I added my algorithm to `DIRECTORY.md` with the correct link.
- [x] I checked `COUNTRIBUTING.md` and my code follows its guidelines.

Please make sure that if there is a test that takes too long to run ( > 300ms), you `#[ignore]` that or
try to optimize your code or make the test easier to run. We have this rule because we have hundreds of
tests to run; If each one of them took 300ms, we would have to wait for a long time.

### CODE REVIEW

File: src/geometry/ramer_douglas_peucker.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
+use crate::geometry::Point;
+
+pub fn ramer_douglas_peucker(polygon: &[Point], epsilon: f64) -> Vec<Point> {
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:3]: Do you really mean `polygon`? I thought this should work for any _polygonal chain_.

File: src/geometry/ramer_douglas_peucker.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    let mut dmax = 0.0;
+    let mut index = 0;
+    let end = polygon.len() - 1;
+
+    for i in 1..end {
+        let d =
+            perpendicular_distance(polygon[i].clone(), polygon[0].clone(), polygon[end].clone());
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:10]:
```suggestion
        let d = perpendicular_distance(&polygon[i], &polygon[0], &polygon[end]);
```

File: src/geometry/ramer_douglas_peucker.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        if d > dmax {
+            index = i;
+            dmax = d;
+        }
+    }
+
+    if dmax > epsilon {
+        let mut results = ramer_douglas_peucker(&Vec::from(&polygon[..=index]), epsilon);
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:18]:
```suggestion
        let mut results = ramer_douglas_peucker(&polygon[..=index], epsilon);
```

File: src/geometry/ramer_douglas_peucker.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            &Vec::from(&polygon[index..]),
+            epsilon,
+        ));
+        results
+    } else {
+        vec![polygon[0].clone(), polygon[end].clone()]
+    }
+}
+
+fn perpendicular_distance(p: Point, a: Point, b: Point) -> f64 {
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:30]:
```suggestion
fn perpendicular_distance(p: &Point, a: &Point, b: &Point) -> f64 {
```

File: src/geometry/ramer_douglas_peucker.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_perpendicular_distance() {
+        let a = Point::new(0.0, 0.0);
+        let b = Point::new(0.0, 3.0);
+        let p = Point::new(4.0, 0.0);
+        assert_eq!(perpendicular_distance(p, a, b), 4.0);
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:45]:
```suggestion
        assert_eq!(perpendicular_distance(&p, &a, &b), 4.0);
```

File: src/geometry/ramer_douglas_peucker.rs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    #[test]
+    fn test_ramer_douglas_peucker() {
+        // This test doesn't make sense, once you change the epsilon value, the result will change. It is hard to predict the result.
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:50]: ?

File: src/geometry/ramer_douglas_peucker.rs [hunk 7]
```diff
@@ -0,0 +1,73 @@
+use crate::geometry::Point;
+
+pub fn ramer_douglas_peucker(polygon: &[Point], epsilon: f64) -> Vec<Point> {
+    let mut dmax = 0.0;
+    let mut index = 0;
+    let end = polygon.len() - 1;
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:6]: What should happen if the input line it _too short_?

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/geometry/ramer_douglas_peucker.rs (hunk 1 above)
Comment (Sellig6792) [reply src/geometry/ramer_douglas_peucker.rs:3]: Yes it does, you are right, should i name it `points`?

File: src/geometry/ramer_douglas_peucker.rs (hunk 6 above)
Comment (Sellig6792) [reply src/geometry/ramer_douglas_peucker.rs:50]: I just mean that the result changes with the epsilon so you can't just predict the output easily

File: src/geometry/ramer_douglas_peucker.rs [hunk 8]
```diff
@@ -0,0 +1,86 @@
+use crate::geometry::Point;
+
+pub fn ramer_douglas_peucker(points: &[Point], epsilon: f64) -> Vec<Point> {
+    if points.len() < 3 {
+        return points.to_vec();
+    }
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:6]: Please add tests, where you call `ramer_douglas_peucker` with empty, single point and two point inputs.

File: src/geometry/ramer_douglas_peucker.rs [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_perpendicular_distance() {
+        let a = Point::new(0.0, 0.0);
+        let b = Point::new(0.0, 3.0);
+        let p = Point::new(4.0, 0.0);
+        assert_eq!(perpendicular_distance(&p, &a, &b), 4.0);
+    }
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:45]:
```suggestion
    macro_rules! test_perpendicular_distance {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (p, a, b, expected) = $test_case;
                    assert_eq!(perpendicular_distance(&p, &a, &b), expected);
                    assert_eq!(perpendicular_distance(&p, &b, &a), expected);
                }
            )*
        };
    }

    test_perpendicular_distance! {
        basic: (Point::new(4.0, 0.0), Point::new(0.0, 0.0), Point::new(0.0, 3.0), 4.0),
        basic_shifted_1: (Point::new(4.0, 1.0), Point::new(0.0, 1.0), Point::new(0.0, 4.0), 4.0),
        basic_shifted_2: (Point::new(2.0, 1.0), Point::new(-2.0, 1.0), Point::new(-2.0, 4.0), 4.0),
    }
```

File: src/geometry/ramer_douglas_peucker.rs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            g.clone(),
+            h.clone(),
+        ];
+        let epsilon = 0.7;
+        let result = ramer_douglas_peucker(&polygon, epsilon);
+        assert_eq!(result, vec![a, c, e, g, h]);
+    }
+
+    #[test]
+    fn test_polygon_chain() {
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:73]: Please change the name of this test.

**Review by vil02 — CHANGES_REQUESTED** [review]

File: src/geometry/ramer_douglas_peucker.rs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        if d > dmax {
+            index = i;
+            dmax = d;
+        }
+    }
+
+    if dmax > epsilon {
+        let mut results = ramer_douglas_peucker(&points[..=index], epsilon);
+        results.pop();
+        results.extend(ramer_douglas_peucker(&Vec::from(&points[index..]), epsilon));
```
Comment (vil02) [inline src/geometry/ramer_douglas_peucker.rs:22]:
```suggestion
        results.extend(ramer_douglas_peucker(&points[index..], epsilon));
```

**Review by vil02 — APPROVED** [review]
